### PR TITLE
feat(p2p): add reputation management features

### LINKED ISSUE #2384: Add peer reputation management features
### Describe the feature

The features requested here are the following:
* Add a method to request the current reputation of a peer to the network handle and its APIs: `reputation_by_id`
* Add `ReputationChangeKind::Reset` to reset a peer's reputation to 0

If you guys approve this I can create a PR :)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mempirate**
Closes #2384

### CODE REVIEW

File: crates/net/network/src/network.rs [hunk 1]
```diff
@@ -213,6 +214,12 @@ impl Peers for NetworkHandle {
     fn reputation_change(&self, peer_id: PeerId, kind: ReputationChangeKind) {
         self.send_message(NetworkHandleMessage::ReputationChange(peer_id, kind));
     }
+
+    async fn reputation_by_id(&self, peer_id: PeerId) -> Result<Option<Reputation>, NetworkError> {
+        let (tx, rx) = oneshot::channel();
+        let _ = self.manager().send(NetworkHandleMessage::GetReputationById(peer_id, tx));
+        rx.await.map_err(|e| e.into())
```
Comment (mattsse) [inline crates/net/network/src/network.rs:221]:
```suggestion
        Ok(rx.await?)
```

File: crates/net/network/src/peers/manager.rs [hunk 2]
```diff
@@ -321,11 +321,20 @@ impl PeersManager {
...
+        self.peers.get(peer_id).map(|peer| peer.reputation)
+    }
+
     /// Apply the corresponding reputation change to the given peer
     pub(crate) fn apply_reputation_change(&mut self, peer_id: &PeerId, rep: ReputationChangeKind) {
         let reputation_change = self.reputation_weights.change(rep);
         let outcome = if let Some(peer) = self.peers.get_mut(peer_id) {
-            peer.apply_reputation(reputation_change.as_i32())
+            // If reputation change is 0, reset reputation to default
+            if reputation_change.as_i32() == 0 {
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:333]: this should check `if rep.is_reset{reset} else convert weights+apply`

**Review by mattsse — CHANGES_REQUESTED** [review]
smol nit

File: crates/net/network/src/peers/manager.rs [hunk 3]
```diff
@@ -327,14 +327,17 @@ impl PeersManager {
 
     /// Apply the corresponding reputation change to the given peer
     pub(crate) fn apply_reputation_change(&mut self, peer_id: &PeerId, rep: ReputationChangeKind) {
+        // if the reputation change is a reset, reset the reputation and return
+        if rep.is_reset() {
+            if let Some(peer) = self.peers.get_mut(peer_id) {
+                peer.reset_reputation();
+            }
+            return
```
Comment (mattsse) [inline crates/net/network/src/peers/manager.rs:335]: pedantic nit:
we only need to do `if let Some(peer) = self.peers.get_mut(peer_id) {` and can do the if is_reset check in that if branch?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/net/network/src/peers/manager.rs (hunk 3 above)
Comment (jonasbostoen) [reply crates/net/network/src/peers/manager.rs]: Oh yeah mb!

**Review by mattsse — APPROVED** [review]

**Pull request merged**